pkg/cloudprovider/vsphere/loadbalancer/lbprovider.go

```diff
@@ -0,0 +1,210 @@
+/*
+ Copyright 2020 The Kubernetes Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package loadbalancer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/cloud-provider-vsphere/pkg/cloudprovider/vsphere/loadbalancer/config"
+
+	"github.com/pkg/errors"
+	nsxt "github.com/vmware/go-vmware-nsxt"
+
+	corev1 "k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+const (
+	// AnnotLoadBalancerClass is the optional class annotation at the service
+	AnnotLoadBalancerClass = "loadbalancer.vmware.io/class"
+)
+
+var (
+	// AppName is set by the main program to the name of the application
+	AppName string
+	// Version is set by the main program to the version of the application
+	Version string
+)
+
+type lbProvider struct {
+	*lbService
+	classes *loadBalancerClasses
+	keyLock *keyLock
+}
+
+// ClusterName contains the cluster name injected from main
+// TODO cluster name needed for reorg and is currently injected from main
+var ClusterName string
+
+var _ cloudprovider.LoadBalancer = &lbProvider{}
+
+// NewLBProvider creates a load balancer implementation for the given config
+func NewLBProvider(cfg *config.LBConfig) (LoadBalancer, error) {
+	if !cfg.IsEnabled() {
```

Comment:
`cfg.IsEnabled` is checking `len(cfg.LoadBalancerClasses) > 0 || !cfg.LoadBalancer.IsEmpty()`, which seems like a configuration validation. Should this be captured in `cfg.LBConfig.CompleteAndValidate()`?

Reply:
If only one of these cases is given or both, the config is valid. The feature is enabled if at least one such entries are given. This is checked and hidden by this method.